Add Content: Observability for java applications (#69)
elastic/observability-docs: docs/en/observability/monitor-java-app.asciidoc

`src/main/java/de/spinscale/javalin/App.java` file.
+
[source,java]
----
package de.spinscale.javalin;

import io.javalin.Javalin;

public class App {
    public static void main(String[] args) {
        Javalin app = Javalin.create().start(7000);
        app.get("/", ctx -> ctx.result("Appsolutely perfect"));
    }
}
----
+
. Run `./gradlew assemble`
+
This creates a compiled `App.class` file in the `build` directory.
However, there is no way to start the server. Let’s create a jar which
contains our compiled class along with all the required dependencies.
+
. In the `build.gradle` file, edit `plugins` as shown here.
+
[source,gradle]
----
plugins {
  id 'com.github.johnrengelman.shadow' version '6.0.0'
  id 'java'
}
----
+
. Run `./gradlew shadowJar`. This creates a file
`build/libs/javalin-app-all.jar`, which now can be run as shown here.
+
[source,bash]
----
java -jar build/libs/javalin-app-all.jar
----
+
The `shadowJar` plugin requires some information about its main class. Add the
following snippet to the `build.gradle` file.
+
....
jar {
  manifest {
    attributes 'Main-Class': 'de.spinscale.javalin.App'
  }
}

> ....

Reviewer: There are a few code blocks like this that aren't formatted correctly. I'm guessing you just haven't got around to this part of the review yet.

Reviewer 2: Thanks. Updated the code blocks.

`src/main/java/de/spinscale/javalin/App.java` file.
+
[source,java]
----
package de.spinscale.javalin;

import io.javalin.Javalin;

public class App {
    public static void main(String[] args) {
        Javalin app = Javalin.create().start(7000);
        app.get("/", ctx -> ctx.result("Appsolutely perfect"));
    }
}
----
+
. Run `./gradlew assemble`
+
This command compiled the `App.class` file in the `build` directory.
However, there is no way to start the server. Let’s create a jar that
contains our compiled class along with all the required dependencies.
+
. In the `build.gradle` file, edit `plugins` as shown here.
+
[source,gradle]
----
plugins {
  id 'com.github.johnrengelman.shadow' version '6.0.0'
  id 'java'
}
----
+
. Run `./gradlew shadowJar`. This command creates a
`build/libs/javalin-app-all.jar` file.
+
The `shadowJar` plugin requires information about its main class.
. Add the following snippet to the `build.gradle` file.
+
[source,gradle]
----
jar {
  manifest {
    attributes 'Main-Class': 'de.spinscale.javalin.App'
  }
}
----
+
. Rebuild the project and start the server.
+
[source,bash]
----

> java -jar build/libs/javalin-app-all.jar

Author: You should not need to do this after adding the `org.apache.logging.log4j:log4j-slf4j18-impl:2.13.3` dependency. Did you do that?

Regarding your above comment, I fixed this by adding another line to the `build.gradle` file.

+
[source,java]
----
package de.spinscale.javalin;

import io.javalin.http.Context;
import org.junit.jupiter.api.Test;

import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.util.HashMap;

import static de.spinscale.javalin.App.mainHandler;
import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.mock;

public class AppTests {

    final HttpServletRequest req = mock(HttpServletRequest.class);
    final HttpServletResponse res = mock(HttpServletResponse.class);
    final Context ctx = new Context(req, res, new HashMap<>());

    @Test
    public void testMainHandler() throws Exception {
        mainHandler().handle(ctx);

        String response = resultStreamToString(ctx);
        assertThat(response).isEqualTo("Appsolutely perfect");
    }

    private String resultStreamToString(Context ctx) throws IOException {
        final byte[] bytes = ctx.resultStream().readAllBytes();
        return new String(bytes, StandardCharsets.UTF_8);
    }
}
----
+
. After the tests pass, build and package the application.
+
[source,bash]
----

> ./gradle clean check shadowJar

Author: yes, must be `gradlew`!

----
+
. Restart the application and send a request. The logs will be sent to
`/tmp/javalin/app.log`.

[discrete]
==== Install and configure {filebeat}

. To read the log file and send it to {es}, {filebeat} is required. For details, see <<install-filebeat,Install {filebeat}>>.
+
. Use the {filebeat} keystore to store {filebeat-ref}/keystore.html[secure
settings].
+
Let’s store the cloud id in the keystore.
+
[source,bash]
----
echo -n "observability-javalin-app:ZXUtY2VudHJhbC0xLmF3cy5jbG91ZC5lcy5pbyQ4NDU5M2I1YmQzYTY0N2NhYjA2MWQ3NTJhZWFhNWEzYyQzYmQwMWE2OTQ2MmQ0N2ExYjdhYTkwMzI0YjJiOTMyYQ==" | ./filebeat keystore add CLOUD_ID --stdin
----
+
To store logs in {es} with minimal permissions, create an API key to send data from {filebeat} to {ess}.
+
. Log into Kibana as the `elastic` user and select *Management* -> *Dev Tools*. Send this request.
+
[source,console]
----
POST /_security/api_key
{
  "name": "filebeat_javalin-app", 
  "role_descriptors": {
    "filebeat_writer": { 
      "cluster": ["monitor", "read_ilm"],
      "index": [
        {
          "names": ["filebeat-*"],
          "privileges": ["view_index_metadata", "create_doc"]
        }
      ]
    }
  }
}
----
+

> The response contains an `api_key` and an `id` field, which can be stored in the {filebeat} keystore.

Author: ++

[discrete]
== Step 3: View logs in Kibana

. Log into Kibana and select the *Discover* app.
+
There is a summary of the documents at the top, but let’s take a look at a single document.
+
image:./images/monitor-java-app-kibana-single-document.png[Kibana single document view]
+
You can see that a lot more data is indexed than just the event. There is information about
the offset in the file, information about the component shipping the logs, the name of the shipper's
name in the output, and there is a `message` field containing log line contents.
+
You can see there is a flaw in the request logging. If the user agent is `null`,
something other than `null` is returned. Reading our logs is
crucial; however, just indexing them gains us nothing.  To fix this, here is a new request logger.
+
[source,java]
----
Javalin app = Javalin.create(config -> {
    config.requestLogger((ctx, executionTimeMs) -> {
        String userAgent = ctx.userAgent() != null ? ctx.userAgent() : "-";
        logger.info("{} {} {} {} \"{}\" {}",
                ctx.method(), ctx.req.getPathInfo(), ctx.res.getStatus(),
                ctx.req.getRemoteHost(), userAgent, executionTimeMs.longValue());
    });

> });

Reviewer: This code does not fix my "null" problem. I still see this:

```
2020-09-10T22:30:25,436-07:00 [INFO ] de.spinscale.javalin.App This is an informative logging message, user agent [null]
```

available in the `message` field.

[IMPORTANT]
=====
The removal of parts of the original message is debatable. Keeping the
original message around makes a lot of sense to me. With the above example,
debugging might become problematic if parsing the timestamp does not work as
expected.
=====

There is also a slight difference in the parsing of a timestamp as the
go time parser only accepts dots as a separator between seconds and
milliseconds. Still, our default output of the log4j2 is using a comma.

Either one can fix the timestamp in the logging output to look like one
expected from {filebeat}. This results in the following pattern layout.

[source,xml]
----
  <PatternLayout pattern="%d{yyyy-MM-dd'T'HH:mm:ss.SSSZ} [%-5level] %logger{36} %msg%n"/>
----

Fixing the timestamp parsing is another way, as you do not always have
full control over your logs and change their format. Imagine using some
third-party software. For now, this will be good enough.

Restart {filebeat} after the change, and look at what
changed in an indexed JSON document by running this search (and having
another log message indexed).

[source,console]
----

> GET filebeat-7.9.0/_search?filter_path=**._source

Reviewer: @EamonnTP This needs to use the version attribute so it stays updated

      <PatternLayout pattern="%d{yyyy-MM-dd'T'HH:mm:ss.SSSZ} [%level] %logger{36} %msg%n"/>
      <Policies>
        <TimeBasedTriggeringPolicy />
        <SizeBasedTriggeringPolicy size="50 MB"/>
      </Policies>
      <DefaultRolloverStrategy max="20"/>
    </RollingFile>
  </Appenders>

  <Loggers>
    <Root level="error">
      <AppenderRef ref="Console" />
      <AppenderRef ref="JavalinAppLogRolling" />
    </Root>
  </Loggers>
</Configuration>
----

The sample added a `JavalinAppLogRolling` appender to our configuration that
uses the same logging pattern as before, but rolls over if a new day
starts or if the log file has reached 50 megabytes. 

If a new log file is created, older log files are gzipped as well to take less space on disk.
The size of 50 megabytes refers to the unpacked file size, so
the potentially twenty files on disk will be much smaller each.

[discrete]
==== Ingest node

The built-in modules are almost entirely using the
{ref}/ingest.html[Ingest node]
feature of {es} instead of the {beats} processors.

One of the most helpful parts of the ingest pipeline is the ability to debug by using
the {ref}/simulate-pipeline-api.html[Simulate Pipeline API].

> . Let’s write a pipeline that is similar to our {filebeat} processors.

Reviewer: ```suggestion
. Let’s write a pipeline that is similar to our {filebeat} processors. Using the Dev Tools panel in Kibana, run the following:
```

    <TimeBasedTriggeringPolicy />
    <SizeBasedTriggeringPolicy size="50 MB"/>
  </Policies>
  <DefaultRolloverStrategy max="20"/>
</RollingFile>
----
+
When you restart your app, you will see pure JSON written to your
log file. When you trigger an exception, you will see, that the
stack trace is already within your single document. This means the
{filebeat} configuration can become stateless and even more lightweight.
Also, the ingest pipeline on the {es} side can be deleted
again.
+
. You can configure a few
https://github.com/elastic/ecs-logging-java/tree/master/log4j2-ecs-layout[more
parameters] for the `EcsLayout`, but the defaults have been selected wisely. Let’s
fix the {filebeat} configuration:
+
[source,yaml]
----
filebeat.inputs:
- type: log
  enabled: true
  paths:
    - /tmp/javalin/*.log
  json.keys_under_root: true

name: javalin-app-shipper

cloud.id: ${CLOUD_ID}
output.elasticsearch:
  api_key: ${ES_API_KEY}

# ================================= Processors =================================
processors:
  - add_host_metadata: ~
----
+
As you can see, just by writing out logs as JSON, our whole logging
setup got a ton easier, so whenever possible, try to directly

> write your logs as JSON.

Reviewer: Upon rebuilding and rerunning, I get the following error:

```
Brandons-MacBook-Pro-4 elastic-obs-java-app $ java -jar build/libs/javalin-app-all.jar
ERROR StatusLogger Unrecognized format specifier [d]
ERROR StatusLogger Unrecognized conversion specifier [d] starting at position 16 in conversion pattern.
ERROR StatusLogger Unrecognized format specifier [thread]
ERROR StatusLogger Unrecognized conversion specifier [thread] starting at position 25 in conversion pattern.
ERROR StatusLogger Unrecognized format specifier [level]
ERROR StatusLogger Unrecognized conversion specifier [level] starting at position 35 in conversion pattern.
ERROR StatusLogger Unrecognized format specifier [logger]
ERROR StatusLogger Unrecognized conversion specifier [logger] starting at position 47 in conversion pattern.
ERROR StatusLogger Unrecognized format specifier [msg]
ERROR StatusLogger Unrecognized conversion specifier [msg] starting at position 54 in conversion pattern.
ERROR StatusLogger Unrecognized format specifier [n]
ERROR StatusLogger Unrecognized conversion specifier [n] starting at position 56 in conversion pattern.
ERROR StatusLogger Unrecognized format specifier [d]
ERROR StatusLogger Unrecognized conversion specifier [d] starting at position 16 in conversion pattern.
ERROR StatusLogger Unrecognized format specifier [thread]
ERROR StatusLogger Unrecognized conversion specifier [thread] starting at position 25 in conversion pattern.
ERROR StatusLogger Unrecognized format specifier [level]
ERROR StatusLogger Unrecognized conversion specifier [level] starting at position 35 in conversion pattern.
ERROR StatusLogger Unrecognized format specifier [logger]
ERROR StatusLogger Unrecognized conversion specifier [logger] starting at position 47 in conversion pattern.
ERROR StatusLogger Unrecognized format specifier [msg]
ERROR StatusLogger Unrecognized conversion specifier [msg] starting at position 54 in conversion pattern.
ERROR StatusLogger Unrecognized format specifier [n]
ERROR StatusLogger Unrecognized conversion specifier [n] starting at position 56 in conversion pattern.
```

I was able to fix this by adding the following to `build.gradle`:
```
shadowJar {
  transform(com.github.jengelman.gradle.plugins.shadow.transformers.Log4j2PluginsCacheFileTransformer)
}
```

We should add this instruction to the tutorial above.

ways of ingesting metrics (the same applies to {filebeat} with different
types of log files and formats), the Prometheus module needs to be enabled.
+
[source,bash]
----
./metricbeat modules enable prometheus
----
+
Add the Prometheus endpoint to poll.
`./modules.d/prometheus.yml`:
+
[source,yaml]
----
- module: prometheus
  period: 10s
  hosts: ["localhost:7000"]
  metrics_path: /metrics
  username: "metrics"
  password: "secret"
  use_types: true
  rate_counters: true
----
+
To improve security, add the username and the
password to the keystore and refer it here as well.
+
. Start {metricbeat} and verify that the Prometheus events are flowing into
{es}.
+
[source,console]
----
GET metricbeat-7.9.0/_search?filter_path=**.prometheus,hits.total
{
  "query": {
    "term": {
      "event.module": "prometheus"
    }
  }
}
----

[discrete]
== Step 5: View metrics in {kib}

As this is custom data from our Javalin app, there is no pre-defined
dashboard for displaying this data.

Let’s check for the number of logging messages per log level.

[source,console]
----
GET metricbeat-7.9.0/_search
{
  "query": {
    "exists": {
      "field": "prometheus.log4j2_events_total.counter"
    }
  }

> }

Author: 😀

[[monitor-java-app]]
== Monitor a Java application

In this tutorial, you learn how to monitor a Java application using the Elastic
Observability apps: Logs, Metrics, APM, and Uptime.

[discrete]
=== What you'll learn

You will learn how to:

- Create a sample Java application.
- Ingest logs using {filebeat} and view your logs in {kib}.
- Ingest metrics using the https://www.elastic.co/guide/en/beats/metricbeat/current/metricbeat-module-prometheus.html[Metricbeat
Prometheus Module] and view your metrics in {kib}.
- Instrument your application using the https://www.elastic.co/guide/en/apm/agent/java/current/[Elastic APM
Java agent].
- Monitor your services using {heartbeat} and view your uptime data in {kib}.

[discrete]
=== Before you begin

> Create a deployment using our hosted {ess} on https://www.elastic.co/cloud/[Elastic Cloud]. The deployment includes

Reviewer 3: ```suggestion
Create a deployment using our hosted {ess} on https://www.elastic.co/cloud/[{ecloud}]. The deployment includes
```

}

repositories {
  jcenter()
}

dependencies {
  implementation 'io.javalin:javalin:3.9.1'

  testImplementation 'org.junit.jupiter:junit-jupiter-api:5.6.2'
  testRuntimeOnly 'org.junit.jupiter:junit-jupiter-engine:5.6.2'
}

application {
  mainClassName = 'de.spinscale.javalin.App'
}

test {
  useJUnitPlatform()
}
----
+
. Run the following command.
+
[source,bash]
----
echo "rootProject.name = 'javalin-app'" >> settings.gradle

mkdir -p src/main/java/de/spinscale/javalin
mkdir -p src/test/java/de/spinscale/javalin
----
+
. Install the Gradle wrapper. An easy way to install Gradle is to use
https://sdkman.io/[sdkman] and run `sdk install gradle 6.5.1`.
+
. Run `./gradlew clean check`. You should see a
successful build that has nothing built or compiled yet.
+
. To create a Javalin server and its first endpoint (the main endpoint), create the
`src/main/java/de/spinscale/javalin/App.java` file.
+
[source,java]
----
package de.spinscale.javalin;

import io.javalin.Javalin;

public class App {
    public static void main(String[] args) {
        Javalin app = Javalin.create().start(7000);
        app.get("/", ctx -> ctx.result("Appsolutely perfect"));
    }
}
----
+

> . Run `./gradlew assemble`

Reviewer: This part was confusing for me. I got `bash: ./gradlew: No such file or directory` when I ran this. Do we need to add a step higher up that specifies the user should first run something like `gradle init`?

Author: indeed, we need a `gradle wrapper` call higher above that is missing, which would install the wrapper into the project. will add.

+
This command compiled the `App.class` file in the `build` directory.
However, there is no way to start the server. Let’s create a jar which
contains our compiled class along with all the required dependencies.
+
. In the `build.gradle` file, edit `plugins` as shown here.
+
[source,gradle]
----
plugins {
  id 'com.github.johnrengelman.shadow' version '6.0.0'
  id 'java'
}
----
+
. Run `./gradlew shadowJar`. This command creates a

> `build/libs/javalin-app-all.jar` file.

Reviewer: I can't get past this step. I get the following error:

```
Brandons-MacBook-Pro-4 elastic-obs-java-app $ ./gradlew shadowJar

FAILURE: Build failed with an exception.

* Where:
Build file '/Users/brandonmorelli/Documents/GitHub/elastic-obs-java-app/build.gradle' line: 17

* What went wrong:
A problem occurred evaluating root project 'javalin-app'.
> Could not find method application() for arguments [build_954lfwy4urhht8sn6t17frb29$_run_closure3@71d3bfa7] on root project 'javalin-app' of type org.gradle.api.Project.

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org
```

Author: can you run `./gradlew --version` and paste the results here?

Reviewer 2: These are the results when I run `./gradlew --version`

------------------------------------------------------------
Gradle 6.5.1
------------------------------------------------------------

Build time:   2020-06-30 06:32:47 UTC
Revision:     66bc713f7169626a7f0134bf452abde51550ea0a

Kotlin:       1.3.72
Groovy:       2.5.11
Ant:          Apache Ant(TM) version 1.10.7 compiled on September 1 2019
JVM:          14.0.2 (Oracle Corporation 14.0.2+12-46)
OS:           Mac OS X 10.14.5 x86_64

Reviewer: ```
------------------------------------------------------------
Gradle 6.5.1
------------------------------------------------------------

Build time:   2020-06-30 06:32:47 UTC
Revision:     66bc713f7169626a7f0134bf452abde51550ea0a

Kotlin:       1.3.72
Groovy:       2.5.11
Ant:          Apache Ant(TM) version 1.10.7 compiled on September 1 2019
JVM:          14.0.2 (Oracle Corporation 14.0.2+12-46)
OS:           Mac OS X 10.15.6 x86_64
```
Looks to be the same as Eamonn

Author: That looks good. Can you share your project somewhere so I can take a look?

  }
}
----
+
. Rebuild the project and start the server.
+
[source,bash]
----
java -jar build/libs/javalin-app-all.jar
----
+
Opening another terminal and running
`curl localhost:7000` displays an HTTP response.
+
. Test the code. Placing everything into the `main()` method makes it difficult to test
the code. However, a dedicated handler fixes this.
+
Refactor the `App` class.
+
[source,java]
----
package de.spinscale.javalin;

import io.javalin.Javalin;
import io.javalin.http.Handler;

public class App {
    
    public static void main(String[] args) {
        Javalin app = Javalin.create().start(7000);
        app.get("/", mainHandler());
    }

    static Handler mainHandler() {
        return ctx -> ctx.result("Appsolutely perfect");
    }
}
----
+
Add a Mockito and an Assertj dependency to the `build.gradle` file.
+
[source,gradle]
----
dependencies {
  implementation 'io.javalin:javalin:3.9.1'

  testImplementation 'org.mockito:mockito-core:3.3.3'
  testImplementation 'org.assertj:assertj-core:3.16.1'
  testImplementation 'org.junit.jupiter:junit-jupiter-api:5.6.2'
  testRuntimeOnly 'org.junit.jupiter:junit-jupiter-engine:5.6.2'
}
----
+

> Create an `AppTests` class in `src/test/java/de/spinscale/javalin`.

Reviewer: I can't get to this step, but it isn't clear to me that this should be in a file named `AppTests.java`. We should clarify.

Author: ++

+
If you want to see the streaming feature at work, run the following curl request in
a loop while sleeping.
+
[source,bash]
----
while $(sleep 0.7) ; do curl localhost:7000 ; done
----
+
. To view a continuous stream of log messages, click *Stream live*. You can also
highlight specific terms, as shown here.
+
image:./images/monitor-java-app-kibana-streaming.png[Kibana Logs UI Streaming]
+
Looking at one of the documents being indexed, you can see that the log message
is contained in a single field. Verify this by looking at one of those documents.
+
[source,console]
----
GET filebeat-*/_search
{
  "size": 1
}
----
+
Things to note:

* When you compare the `@timestamp` field with the timestamp of the log message, you will
notice that it differs. This means that when filtering based on the
`@timestamp` field, you do not get the results you expect. The current
`@timestamp` field reflects the timestamp of when the event was created within
{filebeat}, not the timestamp of when the log event occurred in the
application.
* The log level cannot be filtered, and the same applies to the class name.

[discrete]

> ==== Structure logs

Reviewer 3: Not sure this really belongs under "View logs in Kibana". I wonder if you need a separate step to contain the info about structuring logs, parsing exceptions, configuring log rotations, ingest node, and writing logs as JSON? Seems like a lot to put under "View logs in Kibana".

Reviewer 2: I agree. I grouped the sections under an additional step: Work with your logs.

            return null;
        }

        @Override
        public String prefix() {
            return "javalin";
        }
    });

    public Micrometer() {
        Metrics.addRegistry(registry);
        new JvmGcMetrics().bindTo(Metrics.globalRegistry);
        new JvmHeapPressureMetrics().bindTo(Metrics.globalRegistry);
        new JvmThreadMetrics().bindTo(Metrics.globalRegistry);
        new JvmCompilationMetrics().bindTo(Metrics.globalRegistry);
        new JvmMemoryMetrics().bindTo(Metrics.globalRegistry);
        new Log4j2Metrics().bindTo(Metrics.globalRegistry);
        new UptimeMetrics().bindTo(Metrics.globalRegistry);
        new FileDescriptorMetrics().bindTo(Metrics.globalRegistry);
        new ProcessorMetrics().bindTo(Metrics.globalRegistry);
    }

    public String scrape() {
        return registry.scrape();
    }
}
----
+
. Restart your app and poll the metrics endpoint.
+
[source,bash]
----
curl localhost:7000/metrics -u metrics:secret
----
+
This returns a line based response with one metric per line. This is the
standard Prometheus format.

[discrete]
==== Install and configure {metricbeat}

. Time to install and configure {metricbeat}.

> For details, see <<install-metricbeat,Install {metricbeat}>>.

Reviewer 3: I think we should provide the instructions inline (as I've suggested for Filebeat) so users don't have to change contexts.

+
. Similar to the {filebeat} setup, run the initial set up of all the dashboards
using the admin user, and then use an API key as well.
+
[source,console]
----
POST /_security/api_key
{
  "name": "metricbeat_javalin-app",
  "role_descriptors": {
    "metricbeat_writer": {
      "cluster": ["monitor", "read_ilm"],
      "index": [
        {
          "names": ["metricbeat-*"],
          "privileges": ["view_index_metadata", "create_doc"]
        }
      ]
    }
  }
}
----
+
Don’t forget to do the initial setup like this.
+
[source,bash]
----

> ./metricbeat setup -e -E 'cloud.id=${CLOUD_ID}' -E 'cloud.auth=elastic:YOUR_SUPER_SECRET_PASS'

Reviewer 3: same comment about command syntax applies here. If we reuse our "widgets" for displaying commands, we'll need to add them for other commands. We could do this after this PR is merged to keep this review clean and move it along.

types of log files and formats), the Prometheus module needs to be enabled.
+
[source,bash]
----
./metricbeat modules enable prometheus
----
+
Add the Prometheus endpoint to poll.
`./modules.d/prometheus.yml`:
+
[source,yaml]
----
- module: prometheus
  period: 10s
  hosts: ["localhost:7000"]
  metrics_path: /metrics
  username: "metrics"
  password: "secret"
  use_types: true
  rate_counters: true
----
+
To improve security, add the username and the
password to the keystore and refer it here as well.
+
. Start {metricbeat} and verify that the Prometheus events are flowing into
{es}.
+
[source,console]
----
GET metricbeat-7.9.0/_search?filter_path=**.prometheus,hits.total
{
  "query": {
    "term": {
      "event.module": "prometheus"
    }
  }
}
----

[discrete]
== Step 5: View metrics in {kib}

As this is custom data from our Javalin app, there is no pre-defined
dashboard for displaying this data.

Let’s check for the number of logging messages per log level.

[source,console]
----
GET metricbeat-7.9.0/_search
{
  "query": {
    "exists": {
      "field": "prometheus.log4j2_events_total.counter"
    }
  }
}
----

Visualize the number of log messages over time, split by the

> log level. Since the Elastic Stack 7.7, there is a new way of creating a

Reviewer 3: I recommend mentioning versions or describing a feature as a "new way of creating" things. This kind of information gets stale very quickly and is likely to stay around longer than you want it to.

visualization called `Lens`.

. Log into {kib} and select *Visualize* -> *Create Visualization*.
+
. Create a line chart and select `metricbeat-*` as the source. 
+
The basic idea is to have a

> https://www.elastic.co/guide/en/elasticsearch/reference/current/search-aggregations-metrics-max-aggregation.html[max

Reviewer 3: ```suggestion
{ref}/search-aggregations-metrics-max-aggregation.html[max
```

aggregation] on the y-axis on the `prometheus.log4j2_events_total.rate`
field, whereas the x-axis, is split by date using a

> https://www.elastic.co/guide/en/elasticsearch/reference/current/search-aggregations-bucket-datehistogram-aggregation.html[date_histogram

Reviewer 3: ```suggestion
{ref}/search-aggregations-bucket-datehistogram-aggregation.html[date_histogram
```

aggregation] on the `@timestamp` field.
+
There is one more split within
each date histogram bucket, split by log level, using a

> https://www.elastic.co/guide/en/elasticsearch/reference/7.8/search-aggregations-bucket-terms-aggregation.html[terms

Reviewer 3: ```suggestion
{ref}/search-aggregations-bucket-terms-aggregation.html[terms
```

Reviewer 3: I haven't tested the link that this resolves to, but I'm assuming you aren't intentionally pointing to 7.8 here. It's best to let the shared attributes file resolve the branch.

aggregation] on the `prometheus.labels.level`, which contains the log
level. Also, increase the size of the log level to six to display
every log level.
+
The final result looks like this.
+
image:./images/monitor-java-app-metrics-kibana-create-visualization-log-rate.png[Date Histogram of
the log rate per log level]

[discrete]
==== Visualize open files over time

The second visualization is to be a check for the number of open
files in our application.

As no one can remember all the field names, let’s take a look at the metrics
output again first.

[source,bash]
----
curl -s localhost:7000/metrics -u metrics:secret | grep ^process
process_files_max_files 10240.0
process_cpu_usage 1.8120711232436825E-4
process_uptime_seconds 72903.726
process_start_time_seconds 1.594048883317E9
process_files_open_files 61.0
----

> Let's look at the `process_files_open_files` metric. This should be a rather static value, that rarely changes.

Reviewer 3: ```suggestion
Let's look at the `process_files_open_files` metric. This should be a rather static value that rarely changes.
```

> If you run an application which stores data within the JVM or opens and closes network

Reviewer 3: ```suggestion
If you run an application that stores data within the JVM, or opens and closes network
```

> sockets, this increases and decreases depending on the load. With a web application,

Reviewer 3: ```suggestion
sockets, this metric increases and decreases depending on the load. With a web application,
```

small web application.

. Run `jps` that will contain your App in the process list.
+
[source,bash]
----
$ jps
14224 Jps
82437 Launcher
82438 App
40895
----
+
. Use `lsof` on that process.
+
[source,bash]
----
$ lsof -p 82438
----
+
You will see more output than just all the files being opened,
as a file is also a TCP connection happening right now.
+
. Add an endpoint to increase the number of open files by
having long-running HTTP connections (each connection is also considered
an open file as it requires a file descriptor), and then run `wrk`
against it.
+
[source,java]
----
final Executor executor = CompletableFuture.delayedExecutor(20, TimeUnit.SECONDS);
app.get("/wait", ctx -> {
    CompletableFuture<String> future = CompletableFuture.supplyAsync(() -> "done", executor);
    ctx.result(future);
});
----
+
Every future gets delayed by 20 seconds, which means that a single HTTP
request stays open for 20 seconds.
+
. Let’s run a `wrk` workload.
+
[source,bash]
----
./wrk -c 100 -t 20 -d 5m http://localhost:7000/wait
----
+
Results show that only twenty requests were sent, which makes sense
given the processing time.
+
Now let’s build a visualization using

> https://www.elastic.co/guide/en/kibana/current/lens.html[Lens], a

Reviewer 3: ```suggestion
{kibana-ref}/lens.html[Lens], a
```

> relatively new tool in Kibana which makes it easier to create

Reviewer 3: ```suggestion
relatively new tool in {kib} which makes it easier to create
```

visualizations.
+
image:./images/monitor-java-app-metrics-kibana-create-visualization-open-files.png[Lens
visualization]
+
. Below `Add filter`, select the `metricbeat-*` index
pattern. This will likely use
`filebeat-*` as the default.
+
The x-axis uses the `@timestamp` field -
which in turn will create a `date_histogram` aggregation again. The
y-axis should not be the document count, as that one will always be
stable, but the maximum value of the documents in the buckets. Click on the right of
the field name on the y-axis and select `Max`. This gives you a similar
visualization than shown, with a peak where you ran the `wrk` command above.
+

> . Now let's have a look at the Metrics app in {kib}. Select *Observability* -> *Metrics*.

Reviewer 3: ```suggestion
. Now let's have a look at the {metrics-app} in {kib}. Select *Observability* -> *Metrics*.
```

+
You will only see data from a single shipper. Still, the moment
you are running several services and the ability to group this per
Kubernetes pod or host enables you to spot hosts with elevated
CPU or memory consumption. 
+
. Click *Metrics Explorer*, you can
start exploring your data for specific hosts or the CPU usage across your
nodes.
+
image:./images/monitor-java-app-metrics-ui-prometheus-event-counter.png[Metrics UI Log
Counter]
+
This is an area chart of the total events counter the Javalin app emits.
It’s rising because there is a component
polling an endpoint that in turn produces another log message. The
steeper peek was due to sending more requests. But where is the sudden
drop-off coming from? A JVM restart. As those metrics are not
persisted, they reset on a JVM restart. With that in mind, it’s
often better to log the `rate` instead of the `counter` field.

[discrete]
== Step 6: Instrument the application

The third piece of Observability is Application Performance Management (APM).
An APM setup consists of an APM server which accepts the data (and is

> already running within our Elastic Cloud setup) and an agent delivering

Reviewer 3: ```suggestion
already running within our {ecloud} setup) and an agent delivering
```

the data to the server.

> The agent has two tasks: instrumenting the java application to

Reviewer 3: ```suggestion
The agent has two tasks: instrumenting the Java application to
```

> extract application performance information and sending that data to the APM server.

Reviewer 3: ```suggestion
extract application performance information and sending that data to the APM Server.
```

[discrete]
==== Data Model

Despite a heavily fragmented APM landscape, the terminology roughly is
the same everywhere. The two most important terms are *Spans* and
*Transactions*.

A transaction encapsulates a series of spans, which in turn contain
information about the execution of a piece of code. Let’s take a look at
this screenshot from the Kibana APM UI.

image:./images/monitor-java-app-apm-transactions.png[A transaction with spans]

This is a Spring Boot application. The
`UserProfileController.showProfile()` is called, which is marked as the
transaction. There are two spans within. First, a request is sent to
{es} using the {es} REST client, and after the
response is rendered using Thymeleaf. The request to {es} is faster than the
rendering in this case.

The Java APM agent can instrument specific frameworks
automatically. Spring and Spring Boot are supported well, and the
above data was created by just adding the agent to the Spring Boot
application without doing any configuration.

Also, there are more agents than just the Java one. There are
agents for Go, .NET, Node, Python, Ruby and the browser (RUM). Agents
keep getting added so you may want to check the

> https://www.elastic.co/guide/en/apm/agent/index.html[APM agent

Reviewer 3: ```suggestion
{apm-agents-ref}/index.html[APM agent
```

documentation].

[discrete]
==== Add the APM agent to your code

You have two options to add Java agent instrumentation to your
application.

First, you can add the agent via a parameter, when calling the `java`
binary. This way, it does not interfere with the packaging of the
application. This mechanism instruments the application when starting
up.

First, download the agent, you can check
https://search.maven.org/search?q=g:co.elastic.apm%20AND%20a:elastic-apm-agent[for
the most recent version].

[source,bash]
----
wget https://repo1.maven.org/maven2/co/elastic/apm/elastic-apm-agent/1.17.0/elastic-apm-agent-1.17.0.jar
----

Specify the agent on startup as well as the configuration
parameters of where to send the APM data to. Before starting the Java application,
let’s get an API key for our APM server running in Elastic Cloud.

> When you check your deployment in Elastic Cloud and click on `APM` on

Reviewer 3: ```suggestion
When you check your deployment in {ecloud} and click on `APM` on
```

the left, you will see the `APM Server Secret Token`, which you can use.
Also you can copy the APM endpoint URL from there.

[source,bash]
----
java -javaagent:/path/to/elastic-apm-agent-1.17.0.jar\
  -Delastic.apm.service_name=javalin-app \
  -Delastic.apm.application_packages=de.spinscale.javalin \
  -Delastic.apm.server_urls=$APM_ENDPOINT_URL \
  -Delastic.apm.secret_token=PqWTHGtHZS2i0ZuBol \
  -jar build/libs/javalin-app-all.jar
----

You could now go ahead and open up the APM UI and you should see the
data flowing in.

[discrete]
==== Automatic attachment

If you do not want to change the startup options of your application,
the standalone agent allows you to attach to running JVMs on a host.

This requires you to download the standalone jar. You can find the link
on the

> https://www.elastic.co/guide/en/apm/agent/java/current/setup-attach-cli.html[official

Reviewer 3: ```suggestion
{apm-java-ref}/setup-attach-cli.html[official
```

docs].

To list your locally running java application, you can run

[source,bash]
----
java -jar /path/to/apm-agent-attach-1.17.0-standalone.jar --list
----

> As I usually run more than a single java app on my system, I specify the

Reviewer 3: ```suggestion
As I usually run more than a single Java app on my system, I specify the
```

[source,bash]
----
java -jar /tmp/apm-agent-attach-1.17.0-standalone.jar --pid 30730 \
  --config service_name=javalin-app \
  --config application_packages=de.spinscale.javalin \
  --config server_urls=$APM_ENDPOINT_URL \
  --config secret_token=PqWTHGtHZS2i0ZuBol
----

This above message will return something like this:

[source,text]
----
2020-07-10 15:04:48.144  INFO Attaching the Elastic APM agent to 30730
2020-07-10 15:04:49.649  INFO Done
----

So now the agent was attached to a running application with a special
configuration.

While both of the first two possibilities work, you can also use the third
one: using the APM agent as a direct dependency. This will allow to write
custom spans and transactions within our application.

[discrete]
==== Programmatic setup

. Add the java agent dependency.
+
[source,gradle]
----
dependencies {
  ... 
  implementation 'co.elastic.apm:apm-agent-attach:1.17.0'
  ...
}
----
+
. Instrument the application right at the start in our `main()` method.
+
[source,java]
----
public static void main(String[] args) {
    ElasticApmAttacher.attach();
    ...
}
----
+
We did not configure any endpoint or API tokens yet. While the

> https://www.elastic.co/guide/en/apm/agent/java/current/setup-attach-api.html#setup-attach-api-configuration[documentation]

Reviewer 3: ```suggestion
{apm-java-ref}/setup-attach-api.html#setup-attach-api-configuration[documentation]
```

> recommends using the `src/main/resources/elasticapm.properties` file, I

Reviewer 3: Curious how other writers feel about the use of "I" here. I'm OK with it.

Reviewer: In a tutorial, I agree.

UI.
+
In the overview, you can see that most time is spent in the HTTP
client, which is not too surprising.
+
image:./images/monitor-java-app-apm-ui-javalin-wttr-1.png[Overview]
+
Our transactions for the `/weather/Munich` now contains a span, that
shows how much time is spent on retrieving the weather data. Because
the HTTP client is instrumented automatically, there is no need to do
anything.
+
image:./images/monitor-java-app-apm-ui-javalin-wttr-2.png[Transaction with span]
+
If the `city` parameter if that URL is of high
cardinality, this will result in a high amount of URLs mentioned instead of
the generic endpoint. The solution to this is to fix the transaction naming to this:
+
[source,java]
----
// better transaction names for each URL
app.before(ctx -> ElasticApm.currentTransaction().setName(ctx.method() + " " + ctx.matchedPath()));
----

[discrete]
==== Method tracing via agent configuration

Instead of writing code to trace methods, you can also configure the
agent to do this. Let’s try to figure out if
logging is a bottleneck for our application and trace the request logger
statements we added earlier.

The agent can

> https://www.elastic.co/guide/en/apm/agent/java/current/config-core.html#config-trace-methods[trace

Reviewer 3: ```suggestion
{apm-java-ref}/config-core.html#config-trace-methods[trace
```

        logger.info("{} {} {} {} \"{}\" {}",
                ctx.method(), ctx.req.getPathInfo(), ctx.res.getStatus(),
                ctx.req.getRemoteHost(), userAgent, executionTimeMs.longValue());
    }
}
----
+
. Fix the call in our `App` class.
+
[source,java]
----
config.requestLogger(new Log4j2RequestLogger());
----
+
. Restart your app, and see how much time your logging takes.
+
image:./images/monitor-java-app-apm-ui-logging-trace.png[Logging caller trace]
+
The request logger takes roughly 400 microseconds. The whole request takes
about 1.3 milliseconds. Approximately a third of the processing of our request goes into logging.
+
If you are on the quest for a faster service, you may want to
rethink logging. However, this logging happens after the result is written
to the client, so while the total processing time increases with logging,
responding to the client does not (closing the connection however might
be). Also note that these tests were conducted without a proper warm-up. I
assume that after appropriate JVM warm-up, you will have much faster processing of
requests.

[discrete]
==== Automatic profiling of inferred spans

> Once you have a bigger application with more code paths than our sample

Reviewer 3: ```suggestion
When you have a bigger application with more code paths than our sample
```

Reviewer 3: General style advice is to avoid using "once" unless you mean "one time."

app, you can try to enable the

> https://www.elastic.co/guide/en/apm/agent/java/current/config-profiling.html#config-profiling-inferred-spans-enabled[automatic

Reviewer 3: ```suggestion
{apm-java-ref}/config-profiling.html#config-profiling-inferred-spans-enabled[automatic
```

profiling of inferred spans] by setting the following.

[source,bash]
----
ELASTIC_APM_PROFILING_INFERRED_SPANS_ENABLED=true
----

This mechanism uses the
https://github.com/jvm-profiling-tools/async-profiler[async profiler] to
create spans without you having to instrument anything allowing you to
find bottlenecks faster.

[discrete]
==== Log correlation

By configuring the agent, let's ease the

> https://www.elastic.co/guide/en/apm/agent/java/current/config-logging.html#config-enable-log-correlation[correlation

Reviewer 3: ```suggestion
{apm-java-ref}/config-logging.html#config-enable-log-correlation[correlation
```

of logs] by adding the transaction ids to our logs.

[source,bash]
----
ELASTIC_APM_ENABLE_LOG_CORRELATION=true
----

You can now check the generated log files that are sent

> to {es} via Filebeat. An entry looks like this.

Reviewer 3: ```suggestion
to {es} via {filebeat}. An entry looks like this.
```

[source,json]
----
{
  "@timestamp": "2020-07-13T12:03:22.491Z",
  "log.level": "INFO",
  "message": "GET / 200 0:0:0:0:0:0:0:1 \"curl/7.64.1\" 0",
  "service.name": "my-javalin-app",
  "event.dataset": "my-javalin-app.log",
  "process.thread.name": "qtp34871826-36",
  "log.logger": "de.spinscale.javalin.Log4j2RequestLogger",
  "trace.id": "ed735860ec0cd3ee3bdf80ed7ea47afb",
  "transaction.id": "8af7dff698937dc5"
}
----

Having the `trace.id` and `transaction.id` added, in the case of an error
you will get an `error.id` field.

[IMPORTANT]
=====
We have not covered the

> https://www.elastic.co/guide/en/apm/agent/java/current/opentracing-bridge.html[Elastic

Reviewer 3: ```suggestion
{apm-java-ref}/opentracing-bridge.html[Elastic
```

APM OpenTracing bridge] or looked into the

> https://www.elastic.co/guide/en/apm/agent/java/current/metrics.html[additional

Reviewer 3: ```suggestion
{apm-java-ref}/metrics.html[additional
```

=====

[discrete]
=== Step 7: Ingest Uptime data

There are some basic monitoring capabilities in our application so far. We index
logs (with traces), we index metrics, and we even can look in our app to
figure out single performance bottlenecks thanks to APM. However, there is
still one weak spot. Everything done so far was within the application, and all
the users are reaching the application from the internet.

How about checking if our users have the same experience that our APM data
is suggesting to us. Imagine having a lagging load balancer fronting your app,
that costs you an additional 50ms per request. That would be devastating. Or
TLS negotiation being costly. Even though none of those external events is your fault, you will
still be impacted by this and should try to mitigate those. This means you
need to know about them first.

https://www.elastic.co/uptime-monitoring[Uptime] not only enables you to monitor the availability
of a service, but also graph latencies over time, and get notified about expiring TLS certificates.

[discrete]
==== Setup

Let’s first download {heartbeat} (the polling component) and configure it

> to check for our application. For details, see <<install-heartbeat,Install {heartbeat}>>.

Reviewer 3: Same comment as earlier. I would prefer that this document be self-contained. We can add related links for more info, but shouldn't send users out to a separate task flow to complete these steps.

You can see the execution for the last
scheduled checks, but the duration for each check might be more interesting.
You can see if the latency for one of your checks is going up. 

The interesting part is the world map at the top. You can specify in the
configuration where the check originated, which in this case was in Munich in Europe.
By configuring several {heartbeat}s running across the world, you
can compare latencies and figure out which data center you need to run your
application to be next to your users.

The duration of the monitor is in the low milliseconds, as it is really
fast. Check the monitor for the `httpbin.org` endpoint, and you will see a
much higher duration. In this case, it is about 400ms for each request. This
is not too surprising, because the endpoint is not nearby and you need to
initiate a TLS connection for every request, which is costly.

Do not underestimate the importance of this kind of
monitoring. Also, consider this just the beginning as the next step is
to have synthetics that monitor the correct behavior of your
application, for example, to ensure that your checkout process works all

> the time.

Reviewer 3: Can you add a "What's next" section that points users to the next great thing they want to do?

basic `metricbeat.yaml`.
+
[source,yaml]
----
metricbeat.config.modules:
  path: ${path.config}/modules.d/*.yml
  reload.enabled: false

name: javalin-metrics-shipper

cloud.id: ${CLOUD_ID}
output.elasticsearch:
  api_key: ${ES_API_KEY}

processors:
  - add_host_metadata: ~
  - add_cloud_metadata: ~
  - add_docker_metadata: ~
  - add_kubernetes_metadata: ~
----
+
As {metricbeat} supports dozens of modules, which in turn are different
ways of ingesting metrics (the same applies to {filebeat} with different
types of log files and formats), the Prometheus module needs to be enabled.
+
[source,bash]
----
./metricbeat modules enable prometheus
----
+
Add the Prometheus endpoint to poll in `./modules.d/prometheus.yml`:
+
[source,yaml]
----
- module: prometheus
  period: 10s
  hosts: ["localhost:7000"]
  metrics_path: /metrics
  username: "metrics"
  password: "secret"
  use_types: true
  rate_counters: true
----
+
To improve security, you should add the username and the
password to the keystore and refer to both in the configuration.
+
. Start {metricbeat} and verify that the Prometheus events are flowing into
{es}.
+
[source,console]
----

> GET metricbeat-*/_search?filter_path=**.prometheus,hits.total

Reviewer: Ahhh, a much better idea to use `*`!
